torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,172 @@ def cached():
+                                     "(list, tuple...). "
+                                     f"Got a sequence with an element {originali.__class__.__name__}.")
+
+                # If they require grad we expect the user to add the new parametesr to the optimizer
+                # after registering the parametrization
+                if requires_grad:
+                    originali = Parameter(originali)
+                _register_parameter_or_buffer(self, f"original{i}", originali)
+
+        # Consistency checks:
+        # Since f : A -> B, right_inverse : B -> A, Z and original should live in B
+        # Z = forward(right_inverse(original))
+        Z = self()
+        if Z.dtype != original.dtype:
+            raise ValueError(
+                "Registering a parametrization may not change the dtype of the tensor.\n"
+                f"unparametrized dtype: {original.dtype}\n"
+                f"parametrized dtype: {Z.dtype}"
+            )
+        if Z.shape != original.shape:
+            raise ValueError(
+                "Registering a parametrization may not change the shape of the tensor.\n"
+                f"unarametrized shape: {original.shape}\n"
+                f"parametrized shape: {Z.shape}"
+            )
 
-        It calls the methods ``right_inverse`` (see :func:`register_parametrization`)
-        of the parametrizations in the inverse order that they have been registered.
-        Then, it assigns the result to ``self.original``.
+    def right_inverse(self, value: Tensor) -> None:
+        r"""Calls the methods ``right_inverse`` (see :func:`register_parametrization`)
+        of the parametrizations in the inverse order they were registered in.
+        Then, it stores the result in ``self.original`` if ``right_inverse`` outputs one tensor
+        or in ``self.original0``, ``self.original1``, ... if it outputs several.
 
         Args:
             value (Tensor): Value to which initialize the module
-
-        Raises:
-            RuntimeError: if any of the parametrizations do not implement a ``right_inverse`` method
         """
         with torch.no_grad():
             # See https://github.com/pytorch/pytorch/issues/53103
             for module in reversed(self):  # type: ignore[call-overload]
                 if hasattr(module, "right_inverse"):
                     value = module.right_inverse(value)
-                else:
-                    raise RuntimeError(
-                        "The parametrization '{}' does not implement a 'right_inverse' method. "
-                        "Assigning to a parametrized tensor is only possible when all the parametrizations "
-                        "implement a 'right_inverse' method.".format(module.__class__.__name__)
-                    )
-            self.original.copy_(value)
+                # else we assume that right_inverse is the identity
+            if self.ntensors == 1:
+                # We know that the result is going to have the same dtype
```

Comment:
I'm not sure we can make that assumption here.
You most likely want to double check things here, especially below for the number of Tensors it returns when a Sequence is involved.